File: ontology/generatesubsets.php
<?php

require('../../php-lib/rdfapi.php');
require('../../arc2/ARC2.php');
$odir = "sio/release/";
$cdir = getcwd();
$sio_file = $cdir . "/sio.owl";
$parser = ARC2::getRDFParser();
$parser->parse('file://'.$sio_file);
$triples = $parser->getTriples();
$index = ARC2::getSimpleIndex($triples, false);

$ns = array( 
	'rdf' => 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
	'rdfs' => 'http://www.w3.org/2000/01/rdf-schema#',
	'owl' => 'http://www.w3.org/2002/07/owl#',
	'dct' => 'http://purl.org/dc/terms/',
	'sio' => 'http://semanticscience.org/resource/',
	'dc' => "http://purl.org/dc/elements/1.1/",
	'vann' => "http://purl.org/vocab/vann/",
	'cito' => "http://purl.org/spar/cito/",
	'protege' => "http://protege.stanford.edu/plugins/owl/protege#",
	'xsd' => "http://www.w3.org/2001/XMLSchema#"
);

$equiv = array(
	$ns['owl'].'Class' => $ns['owl'].'EquivalentClass',
	$ns['owl'].'ObjectProperty' => $ns['owl'].'EquivalentProperty',
	$ns['owl'].'DatatypeProperty' => $ns['owl'].'EquivalentProperty'
);

$sio = 'http://semanticscience.org/ontology/sio.owl';
$o = $index[$sio]['http://www.w3.org/2002/07/owl#versionInfo'][0]; 
$sio_version = $o['value'];
$version_iri = "http://semanticscience.org/ontology/sio/v".$sio_version."/sio-release.owl";

function writeGraph($filepath, $myindex)
{
  global $ns;
  $conf = array('ns' => $ns);
  $ser =  ARC2::getRDFXMLSerializer($conf);
  file_put_contents($filepath, $ser->getSerializedIndex($myindex));
}

function getParents($child, &$subClassOf)
{
	$parents = null;
	if(isset($subClassOf[$child])) {		
		foreach($subClassOf[$child] AS $parent) {
			$parents[] = $parent;
			$p = getParents($parent, $subClassOf);
			if(is_array($p)) $parents = array_merge($parents, $p);
		}
	}
	return $parents;
}
function getChildren($parent, &$superClassOf)
{
	$children = null;
	if(isset($superClassOf[$parent])) {
		foreach($superClassOf[$parent] AS $child) {
			$children[] = $child;
			$c = getChildren($child,$superClassOf);
			if(is_array($c)) $children = array_merge($children, $c);
		}
	}
	return $children;
}

// first identify relevant terms
foreach($index AS $s => $p_list) {
	foreach($p_list AS $p => $o_list) {
		// generate the label associative array
		if($p == "http://www.w3.org/2000/01/rdf-schema#label") {
			$label = $o_list[0]['value'];
			
			# camel case labels
			$camel_label_local_part = ucwords($label);
			$camel_label_local_part = str_replace(array(" ",'(',')'), "",$camel_label_local_part);
			foreach($p_list[$ns['rdf'].'type'] AS $p) {
				if($p['value'] == $ns['owl'].'ObjectProperty' || $p['value'] == $ns['owl'].'DatatypeProperty') {
					$camel_label_local_part = lcfirst($camel_label_local_part);
					break;
				}
			}
			$camel_label_uri = "http://semanticscience.org/resource/".urlencode($camel_label_local_part);
			$camel_labels[$s] = $camel_label_uri;

			# dash labels
			$dash_label = str_replace(" ","-", $label);
			$dash_label_uri = "http://semanticscience.org/resource/".urlencode($dash_label);
			$dash_labels[$s] = $dash_label_uri;
		}
		
		// get the subclass relationships for transitive closure
		if($p == $ns['rdfs'].'subClassOf' || $p == $ns['rdfs'].'subPropertyOf') {
			foreach($o_list AS $o) {
				if($o['type'] == 'uri' && $s != $o['value']) {
					$subClassOf[$s][] = $o['value'];
					$superClassOf[$o['value']][] = $s;
				}
			}
		}
		
		// generate the subset array
		if($p == "http://semanticscience.org/resource/subset") {
			foreach($o_list AS $o) {
				$subsets[$s][] = $o['value'];
			}
		}
		
		// generate the relations list
		if($p == $ns['rdf'].'type') {
			foreach($o_list AS $o) {
				if($o['value'] == $ns['owl'].'ObjectProperty' || 
				$o['value'] == $ns['owl'].'DatatypeProperty') {
					$relations[$s] = '';
				}
			}
		}
	}
}

// build up the list of terms required for transitive closures
$tree = array();
foreach($index AS $s => $p_list) {
	// subsets
	if(isset($subsets[$s])) {
		foreach($subsets[$s] AS $i => $subset) {
			// need to exclude rdfs:subClassOf and rdfs:subPropertyOf
			if(strstr($subset,"++")) {
				// generate the transitive closure up and down
				$new_subset_name = substr($subset,0,-2);
				$tree[$new_subset_name]['include'][] = $s;
				$children = getChildren($s, $superClassOf);
				if(is_array($children)) 
					$tree[$new_subset_name]['include'] = array_merge($tree[$new_subset_name]['include'], $children);
					
				$parents = getParents($s, $subClassOf);
				if(is_array($parents)) 
					$tree[$new_subset_name]['include'] = array_merge($tree[$new_subset_name]['include'], $parents);
				unset($subsets[$s][$i]);
			} else if(strstr($subset,"+")) {			
				$new_subset_name = substr($subset,0,-1);
				$tree[$new_subset_name]['include'][] = $s;
				// need to generate transitive closure
				$children = getChildren($s, $superClassOf);
				if(is_array($children)) 
					$tree[$new_subset_name]['include'] = array_merge($tree[$new_subset_name]['include'], $children);
				unset($subsets[$s][$i]);
			}
			if(strstr($subset,"-")) {
				$new_subset_name = substr($subset,0,-1);
				$tree[$new_subset_name]['exclude'][] = $s;
				// need to generate transitive closure
				$children = getChildren($s, $superClassOf);
				if(is_array($children)) 
					$tree[$new_subset_name]['exclude'] = array_merge($tree[$new_subset_name]['exclude'], $children);
				unset($subsets[$s][$i]);				
			}
		}
	}
}

// populate the per subject uri list
foreach($tree AS $subset => $ie_list) {
	// intersect the include + exclude
	if(isset($ie_list['exclude'])) {
		foreach($ie_list['exclude'] AS $exc) {
			foreach($ie_list['include'] AS $i => $inc) {
				if($exc == $inc) {
					unset($ie_list['include'][$i]);
					break;
				}
			}
		}
	}

	foreach($ie_list['include'] AS $s) {
		$subsets[$s][] = $subset;
	}
}


// now select the subclass/subproperty attributes where appropriate
foreach($index AS $s => $p_list) {
	// subsets
	if(isset($subsets[$s])) {
		foreach($subsets[$s] AS $subset) {		
			if($subset == "sadi") {
				// flat list
				$myplist = $p_list;
				if(isset($myplist[$ns['rdfs'].'subClassOf'])) 
					unset($myplist[$ns['rdfs'].'subClassOf']);
				if(isset($myplist[$ns['rdfs'].'subPropertyOf'])) 
					unset($myplist[$ns['rdfs'].'subPropertyOf']);
				$indexes[$subset][$s] = $myplist;
			} else {
				// only keep uri references to those in the subset
				$myplist = $p_list;
				foreach($myplist AS $p => $o_list) {
					foreach($o_list AS $i => $o) {
						$exclude = false;
						if($o['type'] == 'bnode') $exclude = true;
						if($o['type'] == 'uri' && !strstr($o['value'],"owl")) {
							if(isset($subsets[$o['value']])) {
								$exclude = true;
								foreach($subsets[$o['value']] AS $subs) {
									if($subs == $subset) {$exclude = false;break;}
								}
							}
						}
						if($subset=="relations" && 
							($p == $ns['rdfs'].'domain' || $p == $ns['rdfs'].'range')) {
								$exclude = true;
						}
						if($exclude) unset($myplist[$p][$i]);
					}
				}
				$indexes[$subset][$s] = $myplist;
			}
		}
	}
}

foreach($camel_labels AS $s => $uri) {
	$equiv_relation = null;
	$types = $index[$s][ $ns['rdf'].'type'];
	foreach($types AS $o) {
		if(isset($equiv[ $o['value'] ])) {
			$equiv_relation = $equiv[ $o['value'] ];
			$o['value'] = $uri;
			$indexes['equivs'][$s][$equiv_relation][] = $o;
			
			$o['value'] = $s;
			$indexes['equivs'][$uri][$equiv_relation][] = $o;
		}
	}
}

foreach($dash_labels AS $s => $uri) {
	$equiv_relation = null;
	$types = $index[$s][ $ns['rdf'].'type'];
	foreach($types AS $o) {
		if(isset($equiv[ $o['value'] ])) {
			$equiv_relation = $equiv[ $o['value'] ];
			$o['value'] = $uri;
			$indexes['equivs'][$s][$equiv_relation][] = $o;
			
			$o['value'] = $s;
			$indexes['equivs'][$uri][$equiv_relation][] = $o;
		}
	}
}

foreach($index AS $s => $p_list) {	
	// labels document
	if(isset($camel_labels[$s])) $s_uri = $camel_labels[$s];
	else $s_uri = $s;
	
	foreach($p_list AS $p => $o_list) {
		if(isset($camel_labels[$p])) $p_uri = $camel_labels[$p];
		else $p_uri = $p;
		
		foreach($o_list AS $o) {
			if(isset($camel_labels[$o['value']])) {
				$o['value'] = $camel_labels[$o['value']];
			} 
			$indexes['camelcase-label'][$s_uri][$p_uri][] = $o;
		}
	}

	// add dc:identifier
	$o = null;
	$o['value'] = substr($s,strrpos($s,"/")+1);
	$o['type'] = 'literal';
	$index[$s][ $ns['dc'].'identifier'][] = $o;

	// add equivs
	if(isset($indexes['equivs'][$s_uri])) {
		foreach($indexes['equivs'][$s_uri] AS $equiv_rel => $arr) {
			foreach($arr AS $o) {
				$indexes['camelcase-label'][$s_uri][$equiv_rel][] = $o;
			}
		}
	}
}

foreach($index AS $s => $p_list) {	
	// labels document
	if(isset($dash_labels[$s])) $s_uri = $dash_labels[$s];
	else $s_uri = $s;
	
	foreach($p_list AS $p => $o_list) {
		if(isset($dash_labels[$p])) $p_uri = $dash_labels[$p];
		else $p_uri = $p;
		
		foreach($o_list AS $o) {
			if(isset($dash_labels[$o['value']])) {
				$o['value'] = $dash_labels[$o['value']];
			} 
			$indexes['dash-labels'][$s_uri][$p_uri][] = $o;
		}
	}
}

$vi = $index[$sio]['http://www.w3.org/2002/07/owl#versionInfo'];
foreach($indexes AS $subset => $ind) {
	unset($myindex);

	// add the version iri
	//unset($index[$sio]['http://www.w3.org/2002/07/owl#versionInfo']);
	$myindex = $ind;

//	$myindex[$sio] = $index[$sio];
	$sio_subset_file = "sio-subset-".$subset.".owl";
	$o = null;
	$o['value'] = $sio_subset_file;
	$o['type'] = 'literal';
	
	$myindex[$sio]['http://www.w3.org/2002/07/owl#versionInfo'][] = $o;
	
	$sio_versioned_uri = "http://semanticscience.org/ontology/sio/v".$sio_version."/".$sio_subset_file;
	$o = null;
	$o['value'] = $sio_versioned_uri;
	$o['type'] = 'uri';
	$myindex[$sio]['http://www.w3.org/2002/07/owl#versionIRI'][] = $o;
	
	// add rdfs:isDefinedBy
	foreach($myindex AS $s => $p_obj) {
		if(!strstr($s,"semanticscience")) continue;
		$o = null;
		$o['value'] = $sio_versioned_uri;
		$o['type'] = 'uri';
		$myindex[$s]['http://www.w3.org/2000/01/rdf-schema#isDefinedBy'][] = $o;
		
		// @todo add http://open.vocab.org/terms/defines
	}
	
	// add the subset to the main one
	$sio_subset_url = "http://semanticscience.org/ontology/".$sio_subset_file;
	$o = null;
	$o['value'] = $sio_subset_url;
	$o['type'] = 'uri';
	$index[$sio]['http://rdfs.org/ns/void#subset'][] = $o;
	
	
	echo "generating $subset".PHP_EOL;
	writeGraph($odir."sio-subset-$subset.owl",$myindex);
}

echo "generating versioned SIO".PHP_EOL;

$index[$sio]['http://www.w3.org/2002/07/owl#versionInfo'] = $vi;
foreach($index AS $s => $p_obj) {
	if(!strstr($s,"semanticscience")) continue;
	$o = null;
	$o['value'] = 'http://semanticscience.org/ontology/sio.owl';
	$o['type'] = 'uri';
	$index[$s]['http://www.w3.org/2000/01/rdf-schema#isDefinedBy'][] = $o;
}
$o['value'] = $version_iri;
$o['type'] = 'uri';
$index[$sio]['http://www.w3.org/2002/07/owl#versionIRI'][] = $o;

$o['value'] = date('Y-m-d'); #DateTime::ISO8601);
$o['type'] = 'literal';
$o['datatype'] = 'http://www.w3.org/2001/XMLSchema#date';
$index[$sio]['http://purl.org/dc/terms/modified'][] = $o;

copy($odir."sio-subset-dash-labels.owl", $odir."sio-subset-labels.owl");
writeGraph($odir."sio-release.owl", $index);

File: scripts/generate_subsets.py
#!/usr/bin/env python3
"""Generate SIO subset OWL files from sio.owl."""

import shutil
from datetime import date
from pathlib import Path
from urllib.parse import quote

from rdflib import BNode, Graph, Literal, Namespace, URIRef
from rdflib.namespace import DC, DCTERMS, OWL, RDF, RDFS, XSD

VOID = Namespace("http://rdfs.org/ns/void#")
SIO_RES = Namespace("http://semanticscience.org/resource/")
SIO_ONT = URIRef("http://semanticscience.org/ontology/sio.owl")
SIO_SUBSET_PROP = SIO_RES.subset

EQUIV_MAP = {
    OWL.Class: OWL.equivalentClass,
    OWL.ObjectProperty: OWL.equivalentProperty,
    OWL.DatatypeProperty: OWL.equivalentProperty,
}

ROOT = Path(__file__).parent.parent
ONTOLOGY_DIR = ROOT / "ontology"
ODIR = ONTOLOGY_DIR / "sio" / "release"


def get_children(node, super_class_of, visited=None):
    if visited is None:
        visited = set()
    result = []
    if node in super_class_of and node not in visited:
        visited.add(node)
        for child in super_class_of[node]:
            result.append(child)
            result.extend(get_children(child, super_class_of, visited))
    return result


def get_parents(node, sub_class_of, visited=None):
    if visited is None:
        visited = set()
    result = []
    if node in sub_class_of and node not in visited:
        visited.add(node)
        for parent in sub_class_of[node]:
            result.append(parent)
            result.extend(get_parents(parent, sub_class_of, visited))
    return result


def make_camel_uri(label, is_property):
    camel = "".join(w.capitalize() for w in label.split())
    camel = camel.replace("(", "").replace(")", "")
    if is_property and camel:
        camel = camel[0].lower() + camel[1:]
    return str(SIO_RES) + quote(camel)


def make_dash_uri(label):
    return str(SIO_RES) + quote(label.replace(" ", "-"))


def main():
    sio_file = ONTOLOGY_DIR / "sio.owl"
    g = Graph()
    g.parse(str(sio_file))

    sio_version = str(g.value(SIO_ONT, OWL.versionInfo))
    version_iri = f"http://semanticscience.org/ontology/sio/v{sio_version}/sio-release.owl"

    camel_labels = {}   # str(subject) -> str(camelCase URI)
    dash_labels = {}    # str(subject) -> str(dash URI)
    sub_class_of = {}   # str(child) -> [str(parent), ...]
    super_class_of = {} # str(parent) -> [str(child), ...]
    subsets = {}        # str(subject) -> [subset_name, ...]

    for s, p, o in g:
        s_str = str(s)

        if p == RDFS.label and isinstance(o, Literal):
            label = str(o)
            types = set(g.objects(s, RDF.type))
            is_prop = OWL.ObjectProperty in types or OWL.DatatypeProperty in types
            camel_labels[s_str] = make_camel_uri(label, is_prop)
            dash_labels[s_str] = make_dash_uri(label)

        if p in (RDFS.subClassOf, RDFS.subPropertyOf) and isinstance(o, URIRef) and s != o:
            sub_class_of.setdefault(s_str, []).append(str(o))
            super_class_of.setdefault(str(o), []).append(s_str)

        if p == SIO_SUBSET_PROP:
            subsets.setdefault(s_str, []).append(str(o))

    # Process transitive-closure annotations (++, +, -)
    tree = {}  # name -> {'include': set, 'exclude': set}

    for s_str, subset_list in subsets.items():
        for subset_val in list(subset_list):
            if "++" in subset_val:
                name = subset_val[:-2]
                ie = tree.setdefault(name, {"include": set(), "exclude": set()})
                ie["include"].add(s_str)
                ie["include"].update(get_children(s_str, super_class_of))
                ie["include"].update(get_parents(s_str, sub_class_of))
                subset_list.remove(subset_val)
            elif "+" in subset_val:
                name = subset_val[:-1]
                ie = tree.setdefault(name, {"include": set(), "exclude": set()})
                ie["include"].add(s_str)
                ie["include"].update(get_children(s_str, super_class_of))
                subset_list.remove(subset_val)

            if "-" in subset_val and subset_val in subset_list:
                name = subset_val[:-1]
                ie = tree.setdefault(name, {"include": set(), "exclude": set()})
                ie["exclude"].add(s_str)
                ie["exclude"].update(get_children(s_str, super_class_of))
                subset_list.remove(subset_val)

    for name, ie in tree.items():
        for s_str in ie["include"] - ie["exclude"]:
            subsets.setdefault(s_str, []).append(name)

    # Build per-subset graphs with filtered triples
    subset_graphs = {}

    for s, p, o in g:
        s_str = str(s)
        if s_str not in subsets:
            continue

        for subset in subsets[s_str]:
            sg = subset_graphs.setdefault(subset, Graph())

            if subset == "sadi":
                if p not in (RDFS.subClassOf, RDFS.subPropertyOf):
                    sg.add((s, p, o))
                continue

            if isinstance(o, BNode):
                continue
            if isinstance(o, URIRef):
                o_str = str(o)
                if "owl" not in o_str and o_str in subsets:
                    if subset not in subsets[o_str]:
                        continue
            if subset == "relations" and p in (RDFS.domain, RDFS.range):
                continue

            sg.add((s, p, o))

    # Build equivalents graph (camelCase <-> dash <-> original URIs)
    equivs_g = Graph()
    for mapping in (camel_labels, dash_labels):
        for s_str, alt_uri in mapping.items():
            s = URIRef(s_str)
            alt = URIRef(alt_uri)
            for rdf_type in g.objects(s, RDF.type):
                if rdf_type in EQUIV_MAP:
                    equiv_rel = EQUIV_MAP[rdf_type]
                    equivs_g.add((s, equiv_rel, alt))
                    equivs_g.add((alt, equiv_rel, s))

    # Add dc:identifier to main graph
    for s in set(g.subjects()):
        if isinstance(s, URIRef):
            local_part = str(s).rsplit("/", 1)[-1]
            g.add((s, DC.identifier, Literal(local_part)))

    # Build camelcase-label graph: remap all URIs to camelCase equivalents
    camel_g = Graph()
    for s, p, o in g:
        s_node = URIRef(camel_labels.get(str(s), str(s))) if isinstance(s, URIRef) else s
        p_node = URIRef(camel_labels.get(str(p), str(p)))
        o_node = URIRef(camel_labels.get(str(o), str(o))) if isinstance(o, URIRef) else o
        camel_g.add((s_node, p_node, o_node))

    for s, p, o in equivs_g:
        s_node = URIRef(camel_labels.get(str(s), str(s))) if isinstance(s, URIRef) else s
        camel_g.add((s_node, p, o))

    # Build dash-labels graph: remap all URIs to dash equivalents
    dash_g = Graph()
    for s, p, o in g:
        s_node = URIRef(dash_labels.get(str(s), str(s))) if isinstance(s, URIRef) else s
        p_node = URIRef(dash_labels.get(str(p), str(p)))
        o_node = URIRef(dash_labels.get(str(o), str(o))) if isinstance(o, URIRef) else o
        dash_g.add((s_node, p_node, o_node))

    subset_graphs["camelcase-label"] = camel_g
    subset_graphs["dash-labels"] = dash_g

    # Collect ontology-level metadata to carry into every subset.
    # versionInfo, versionIRI, and void:subset are subset-specific and handled below.
    SUBSET_OVERRIDE_PREDICATES = {OWL.versionInfo, OWL.versionIRI, VOID.subset}
    ontology_metadata = [
        (p, o)
        for p, o in g.predicate_objects(SIO_ONT)
        if p not in SUBSET_OVERRIDE_PREDICATES
    ]

    # Write subset files
    ODIR.mkdir(parents=True, exist_ok=True)
    for subset_name, sg in subset_graphs.items():
        subset_file = f"sio-subset-{subset_name}.owl"
        versioned_uri = f"http://semanticscience.org/ontology/sio/v{sio_version}/{subset_file}"
        subset_url = f"http://semanticscience.org/ontology/{subset_file}"

        for p, o in ontology_metadata:
            sg.add((SIO_ONT, p, o))
        sg.add((SIO_ONT, OWL.versionInfo, Literal(subset_file)))
        sg.add((SIO_ONT, OWL.versionIRI, URIRef(versioned_uri)))

        for s in set(sg.subjects()):
            if isinstance(s, URIRef) and "semanticscience" in str(s):
                sg.add((s, RDFS.isDefinedBy, URIRef(versioned_uri)))

        g.add((SIO_ONT, VOID.subset, URIRef(subset_url)))

        print(f"generating {subset_name}")
        sg.serialize(str(ODIR / subset_file), format="xml")

    # Write versioned release file
    print("generating versioned SIO")
    for s in set(g.subjects()):
        if isinstance(s, URIRef) and "semanticscience" in str(s):
            g.add((s, RDFS.isDefinedBy, URIRef("http://semanticscience.org/ontology/sio.owl")))

    g.add((SIO_ONT, OWL.versionIRI, URIRef(version_iri)))
    g.add((SIO_ONT, DCTERMS.modified, Literal(date.today().isoformat(), datatype=XSD.date)))

    shutil.copy(ODIR / "sio-subset-dash-labels.owl", ODIR / "sio-subset-labels.owl")
    g.serialize(str(ODIR / "sio-release.owl"), format="xml")


if __name__ == "__main__":
    main()
